tools/lib/bootpool.js

```diff
@@ -0,0 +1,279 @@
+        device: null,           /* Will be filled in if successful. */
+        version: null,          /* Will be filled in if successful. */
+        options: null,          /* Will be filled in if successful. */
+        steps: {                /* Will be filled in if successful. */
+            mounted: false,
+            options_ok: false,
+            marker_file: false
+        },
+        ok: false,
+        message: 'not mounted'
+    };
+
+    dprintf('fetching pool mount status for "%s"\n', mountpoint);
+
+    lib_usbkey.get_mount_info(mountpoint, function mnttab_info(err, mi) {
+        if (err) {
+            callback(new VError(err, 'could not inspect mounted filesystems'),
+                    status);
+            return;
+        }
+
+        /*
+         * Okay, convert our mount information into a JSON object like the
+         * USB one.  Version is always '2' (loader) for ZFS pool boots.
+         * "device" will be the bootfs.
+         */
+
+        if (mi === false) {
+            dprintf('ok, bootfs says nothing is not mounted.\n');
+            callback(null, status);
+            return;
+        }
+
+        mod_assert.strictEqual(mi.mi_mountpoint, mountpoint);
+        status.mountpoint = mountpoint;
+        status.device = mi.mi_special;
+        if (mi.mi_fstype === 'lofs') {
+            /* AHA! We're probably good! */
+            status.version = 2; /* Actually consumed by Triton components! */
+            status.options = {}; /* Filled in as an empty list. */
+            status.steps.mounted = true;
+            status.steps.options_ok = true;
+            /* NOTE: marker filled in below. */
+            status.ok = true;
+            status.message = 'mounted';
+        } else {
+            status.message = 'Something not lofs-mounted on ' + mountpoint;
+        }
+
+        /*
+         * Set marker file attribute regardless of correct or not so
+         * we can avoid callback indentation hell.
+         */
+        lib_usbkey.check_for_marker_file(mountpoint, function (err, marker) {
+            if (err) {
+                    callback(new VError(err, 'failed to locate marker on "%s"',
+                        mountpoint));
+                    return;
+            }
+
+            status.steps.marker_file = marker;
+            callback(null, status);
+        });
+    });
+}
+
+/*
+ * Mount bootfs onto the default mountpoint.
+ */
+function
+ensure_bootfs_mounted(poolname, callback)
+{
+    var args = [];
+    /* XXX KEBE ASKS -- use SMF properties? */
+    var mountpoint = lib_usbkey.DEFAULT_MOUNTPOINT;
+
+    mod_assert.func(callback, 'callback');
+    mod_assert.string(poolname, 'poolname');
+
+    dprintf('ensuring usb key is mounted (no altmountopts for bootfs)...\n');
+
+    args.push('list');
+    args.push('-Ho');
+    args.push('bootfs');
+    args.push(poolname);
+    mod_child.execFile('zpool', args, function (err, stdout, stderr) {
```

Comment:
Similarly here as well (callback naming)

Reply:
Gotcha.

Reply:
Leaving it.